x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/service/ServiceAccountService.java

```diff
@@ -0,0 +1,155 @@
+import org.elasticsearch.xpack.core.security.user.User;
+import org.elasticsearch.xpack.security.authc.service.ServiceAccount.ServiceAccountId;
+import org.elasticsearch.xpack.security.authc.support.SecurityTokenType;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.util.Base64;
+import java.util.Map;
+
+import static org.elasticsearch.xpack.security.authc.service.ElasticServiceAccounts.ACCOUNTS;
+
+public class ServiceAccountService {
+
+    public static final String REALM_TYPE = "service_account";
+    public static final String REALM_NAME = "service_account";
+
+    private static final Logger logger = LogManager.getLogger(ServiceAccountService.class);
+    private static final Version VERSION_MINIMUM = Version.V_8_0_0;
+
+    private final ServiceAccountsCredentialStore serviceAccountsCredentialStore;
+
+    public ServiceAccountService(ServiceAccountsCredentialStore serviceAccountsCredentialStore) {
+        this.serviceAccountsCredentialStore = serviceAccountsCredentialStore;
+    }
+
+    public static boolean isServiceAccount(Authentication authentication) {
+        return REALM_TYPE.equals(authentication.getAuthenticatedBy().getType());
+    }
+
+    // {@link org.elasticsearch.xpack.security.authc.TokenService#extractBearerTokenFromHeader extracted} from an HTTP authorization header.
+    /**
+     * Parses a token object from the content of a {@link ServiceAccountToken#asBearerString()} bearer string}.
+     * This bearer string would typically be
+     *
+     * <p>
+     * <strong>This method does not validate the credential, it simply parses it.</strong>
+     * There is no guarantee that the {@link ServiceAccountToken#getSecret() secret} is valid,
+     * or even that the {@link ServiceAccountToken#getAccountId() account} exists.
+     * </p>
+     * @param token A raw token string (if this is from an HTTP header, then the <code>"Bearer "</code> prefix must be removed before
+     *              calling this method.
+     * @return An unvalidated token object.
+     */
+    public static ServiceAccountToken tryParseToken(SecureString token) {
+        try {
+            if (token == null) {
+                return null;
+            }
+            return doParseToken(token);
+        } catch (IOException e) {
+            logger.debug("Cannot parse possible service account token", e);
+            return null;
+        }
+    }
+
+    public void authenticateWithToken(ServiceAccountToken token, ThreadContext threadContext, String nodeName,
+                                      ActionListener<Authentication> listener) {
+
+        if (ElasticServiceAccounts.NAMESPACE.equals(token.getAccountId().namespace()) == false) {
+            logger.debug("only [{}] service accounts are supported, but received [{}]",
+                ElasticServiceAccounts.NAMESPACE, token.getAccountId().asPrincipal());
+            listener.onResponse(null);
```

Comment:
We could make this `onFailure` instead. Since we have a parsed token, we can assume that the client intended to auth as a service account, and actively fail if we cannot.